packages/flutter_test/lib/src/finders.dart

```diff
@@ -873,112 +1397,214 @@ class _WidgetFinder extends MatchFinder {
-    }
-    return '${descendant.description} that has ancestor(s) with ${ancestor.description}';
+  String describeMatch(Plurality plurality) {
+    return _describeMatch?.call(plurality) ??
+      'matching semantics predicate';
   }
 
   @override
-  Iterable<Element> apply(Iterable<Element> candidates) {
-    final Iterable<Element> descendants = descendant.evaluate();
-    return candidates.where((Element element) => descendants.contains(element));
+  bool matches(SemanticsNode candidate) {
+    return predicate(candidate);
   }
+}
+
+mixin _DescendantFinderMixin<CandidateType> on FinderBase<CandidateType> {
+
+  FinderBase<CandidateType> get ancestor;
+  FinderBase<CandidateType> get descendant;
+  bool get matchRoot;
 
   @override
-  Iterable<Element> get allCandidates {
-    final Iterable<Element> ancestorElements = ancestor.evaluate();
-    final List<Element> candidates = ancestorElements.expand<Element>(
-      (Element element) => collectAllElementsFrom(element, skipOffstage: skipOffstage)
+  String describeMatch(Plurality plurality) {
+    return '${descendant.describeMatch(plurality)} descending from '
+      '${ancestor.describeMatch(plurality)}'
+      '${matchRoot ? ' inclusive' : ''}';
+  }
+
+  @override
+  Iterable<CandidateType> findInCandidates(Iterable<CandidateType> candidates) {
+    final Iterable<CandidateType> descendants = descendant.find();
+    return candidates.where((CandidateType candidate) => descendants.contains(candidate));
+  }
+
+  @override
+  Iterable<CandidateType> get allCandidates {
+    final Iterable<CandidateType> ancestors = ancestor.find();
+    final List<CandidateType> candidates = ancestors.expand<CandidateType>(
+      (CandidateType ancestor) => _collectDescendants(ancestor)
     ).toSet().toList();
     if (matchRoot) {
-      candidates.insertAll(0, ancestorElements);
+      candidates.insertAll(0, ancestors);
     }
     return candidates;
   }
+
+  Iterable<CandidateType> _collectDescendants(CandidateType root);
 }
 
-class _AncestorFinder extends Finder {
-  _AncestorFinder(this.descendant, this.ancestor, { this.matchRoot = false }) : super(skipOffstage: false);
+class _DescendantWidgetFinder extends WidgetFinder
+    with _DescendantFinderMixin<Element> {
+  _DescendantWidgetFinder(
+    this.ancestor,
+    this.descendant, {
+    this.matchRoot = false,
+    super.skipOffstage,
+  });
 
-  final Finder ancestor;
-  final Finder descendant;
+  @override
+  final FinderBase<Element> ancestor;
+  @override
+  final FinderBase<Element> descendant;
+  @override
   final bool matchRoot;
 
   @override
-  String get description {
-    if (matchRoot) {
-      return 'ancestor ${ancestor.description} beginning with ${descendant.description}';
-    }
-    return '${ancestor.description} which is an ancestor of ${descendant.description}';
+  String get description => throw UnimplementedError('This is handled by `_DescendantFinderMixin.describeMatch`');
```

Comment:
probably should just keep the old implementation for full backwards compatibility during the deprecation period.

Reply:
Updated to use `describeMatch`.